examples/browser/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+extern crate midir;
+extern crate console_error_panic_hook;
+extern crate js_sys;
+extern crate web_sys;
+extern crate wasm_bindgen;
+
+use js_sys::{Array};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::{console};
+
+use std::sync::{Arc, Mutex};
+use std::error::Error;
+
+use midir::{MidiInput, Ignore};
+
+pub fn log(s: String) {
+    console::log(&Array::of1(&s.into()));
+}
+
+macro_rules! println {
+    ()              => (log("".to_owned()));
+    ($($arg:tt)*)   => (log(format!($($arg)*)));
+}
+
+#[wasm_bindgen(start)]
+pub fn start() {
+    std::panic::set_hook(Box::new(console_error_panic_hook::hook));
+
+    let token_outer = Arc::new(Mutex::new(None));
+    let token = token_outer.clone();
+    let closure : Closure<dyn FnMut()> = Closure::wrap(Box::new(move ||{
+        if run().unwrap() == true {
+            if let Some(token) = *token.lock().unwrap() {
+                web_sys::window().unwrap().clear_interval_with_handle(token);
+            }
+        }
+    }));
+    *token_outer.lock().unwrap() = web_sys::window().unwrap().set_interval_with_callback_and_timeout_and_arguments_0(
+        closure.as_ref().unchecked_ref(),
+        200,
+    ).ok();
+    closure.forget();
+}
+
+fn run() -> Result<bool, Box<dyn Error>> {
+    let mut midi_in = MidiInput::new("midir reading input")?;
+    midi_in.ignore(Ignore::None);
+
+    // Get an input port (read from console if multiple are available)
```

Comment:
> Wow, this is awesome! Thanks a lot!

My pleasure 👍 

> I have only played a tiny little bit with wasm so far, so there is a lot of things here that I don't have a good understanding for, but I'll try my best.

Feel free to ask questions, I'm happy to chime in.

> One thing that's a bit unfortunate is that I have a branch in development (#38) which is almost finished and ready to be merged, which changes the API. So all backends need to be adapted. Do you think that you can rebase the Web MIDI support on that branch?

Sure!  It's what I get for my PR first, ask questions later style :).  I've subscribed to #38 so I should get a notification for when that lands as well.

> It might even make the code leaner and cleaner because ports are no longer identified by the index and it seems that your `DeviceSet` might become simpler or superfluous then.

Looks like it, yeah, I should be able to simply lookup in the map.

> Concerning the asychronicity:
> The issue seems to be in how to handle `js_sys::Promise`, right

Yes.

> I think the only way to really fix this is to introduce Rust's new `async` into midir's API.

That would be one of the more solid options I think, yeah.

> Which is probably a good idea anyway, but out of scope for this PR. (Can `js_sys::Promise` interoperate with `async`?)

I believe so - I've mucked with `Future`s some, but haven't actually touched `async` yet, since only the former has landed in stable so far.

> You can open a follow-up issue so we don't forget about it, and that also serves as documentation about this "known issue" (i.e. that messages might be lost because the asynchronous port opening has not completed yet).

Sounds good!  https://github.com/Boddlnagg/midir/issues/48

> I have left some more comments inline.

Will address and rebase 👍